### PR TITLE
RavenDB-21926: Experimental strategies to improve allocations and performance

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-21926

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [ ] Moderate 
- [x] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### CODE REVIEW

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (ayende) [inline src/Voron/Data/BTrees/TreeMutableState.cs]: I don't think that this should be 1:1 memory aligned as the persistent struct.
This is meant to allow far greater freedom, and isn't persisted, after all.
Better to let C# to just order this as needed. For example, all the longs aren't 8 bytes aligned, etc.

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (ayende) [inline src/Voron/Data/BTrees/TreeMutableState.cs]: You are trying to get cast identical, but why?
Also, if they are structurally the same, why wouldn't this just be `TreeRootHeader` ?

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (ayende) [inline src/Voron/Data/BTrees/TreeMutableState.cs]: How expensive is this that we need to go to all this trouble?

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/TreeMutableState.cs]: The idea is to allow the JIT to emit vector instructions instead. If we do field by field, then for example a copy would require several sequences of mov to register, mov to location for each field.

Comment (ayende) [reply src/Voron/Data/BTrees/TreeMutableState.cs]: Sure, but that happens _once_ per transaction commit.
Accessing those fields is a lot more common, and happens a lot more frequently

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/TreeMutableState.cs]: They are not structurally same, just structurally compatible. The idea behind the sequence of commits that will comprise this experimental PR is to optimize the hell out of the sections that happen under the lock (even if at the expense of maintainability). Then we can mix and match which ones we are OK to keep, and push them as individual optimizations. For example, in this case I got a very modest 500MB of allocations less (in 36Gb is not that much), but it proved that pushing this type of optimizations are easier (even if high riks) than what I would have expected.

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/TreeMutableState.cs]: The JIT will emit 4 assembler instructions vs 2 per field.

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/TreeMutableState.cs]: https://github.com/ravendb/ravendb/pull/17981#discussion_r1448691298

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (ayende) [reply src/Voron/Data/BTrees/TreeMutableState.cs]: What's the difference between those?
Why not just use `TreeRootHeader` directly?

Comment (redknightlois) [reply src/Voron/Data/BTrees/TreeMutableState.cs]: We need the `IsReadOnly` and `IsModified` fields which are not in the `TreeRootHeader` ... We could technically just store the `TreeRootHeader` in there if that is what you mean. But that would require to modify a million files.

Comment (ayende) [reply src/Voron/Data/BTrees/TreeMutableState.cs]: Can't we add that to the `TreeRootHeader` directly, then just change the type? Or methods

**Review by ayende — APPROVED** [review]

**Comment by redknightlois** [comment]
Requires: https://github.com/ravendb/ravendb/pull/18159. There is no guarantee of working without solving the multiple dispose issue. 

https://issues.hibernatingrhinos.com/issue/RavenDB-22090/Tree.Dispose-being-called-multiple-times-and-Trees-been-used-after-dispose

**Comment by redknightlois** [comment]
Requires: https://github.com/ravendb/ravendb/pull/18106

**Comment by redknightlois** [comment]
Requires https://github.com/ravendb/ravendb/pull/18180 to be able to fully test.

**Comment by redknightlois** [comment]
Pretty sure that I can adjust this a bit more, it seems like there is a lot of contention and that causes lots of excess allocations. 

image

**Comment by arekpalinski** [comment]
Please do the rebase and solve the conflict.

**Comment by arekpalinski** [comment]
Please check failing tests: https://jenkins.hibernatingrhinos.com/job/PR_Tests/10081/

**Comment by redknightlois** [comment]
New approach requires: https://github.com/ravendb/ravendb/pull/18212

**Comment by arekpalinski** [comment]
1.  > New approach requires: #18212

https://github.com/ravendb/ravendb/pull/18212 seems to be effectively the same set of changes as here (all commits from https://github.com/ravendb/ravendb/pull/18212 are here). There is one additional commit here (I believe addressing this comment: https://github.com/ravendb/ravendb/pull/18212#discussion_r1555351495)

Although https://github.com/ravendb/ravendb/pull/18212 has 10 files modified while here we have changes in 54 files. I believe we can close one of the PRs? Which one? Please handle that

2. This one is marked with `DO NOT MERGE` label - is it still the case?

**Comment by arekpalinski** [comment]
test this in debug please

File: src/Voron/Data/BTrees/Tree.MultiTree.cs
Comment (arekpalinski) [inline src/Voron/Data/BTrees/Tree.MultiTree.cs]: TODO?

File: src/Voron/Data/BTrees/Tree.Compressed.cs
Comment (arekpalinski) [inline src/Voron/Data/BTrees/Tree.Compressed.cs:250]: As I understand we need to call `Modify()` explicitly now before doing any modification of the state, right? This construction might be error prone. Can we have a similar concept like we have in `HeaderAccessor.Modify(ModifyHeaderAction modifyAction)`? So here we'd have something like:

```
State.Modify(header => header.NumberOfEntries--);
```

File: src/Voron/Impl/LowLevelTransaction.cs
Comment (arekpalinski) [inline src/Voron/Impl/LowLevelTransaction.cs]: Please wrap it in

```
#if DEBUG
using System.Linq; // Needed in DEBUG
using System.Text; // Needed in DEBUG
#endif

```

File: src/Voron/Data/BTrees/Tree.MultiTree.cs
Comment (ayende) [inline src/Voron/Data/BTrees/Tree.MultiTree.cs]: I... don't like it.

This retruns a `ref` reference, which you are discarding. 

Then you are modifying the `tree.State.Header` - is that the same one? Something else?

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (ayende) [inline src/Voron/Data/BTrees/TreeMutableState.cs:13]: If we have `Modify`, why do we expose this directly?
Wouldn't it be better to create properties that would expose just the right values (typically `RootPageNumber`, I think) that would be elided by the JIT?

File: src/Voron/Data/BTrees/Tree.cs
Comment (ayende) [inline src/Voron/Data/BTrees/Tree.cs:1619]: Why?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/TreeMutableState.cs:13]: It is a `readonly` reference. Most of the time we don't need to modify.

File: src/Voron/Data/BTrees/Tree.Compressed.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/Tree.Compressed.cs:250]: No, this is very low level access and therefore performance sensitive. 

We do that there for 2 reasons:
- We need to first get the lock to be able to execute the action. 
- We almost never have to run that code. 

This way if we call `ref state = ref State.Modify()` serves the dual purpose of, do not allow the cases where we modify the state and not notify of that modification (usual error we had back then in 2014 and took ages to get rid everywhere) and also to provide an alternative way to access the data through the `ref readonly Header` as a direct reference operation (as high performance as a dereference from a register).

File: src/Voron/Data/BTrees/Tree.MultiTree.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/Tree.MultiTree.cs]: Let me remove it and run the slow tests. I marked that because I was not entirely sure if it was needed and forgot to recheck it.

File: src/Voron/Data/BTrees/Tree.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/Tree.cs:1619]: This is actually a special case, because we are going to be renaming the tree and we have to force the state to be considered as Modified even though we are not going to change anything in it. I will add a comment in there.

File: src/Voron/Data/BTrees/Tree.MultiTree.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/Tree.MultiTree.cs]: `tree.State.Header` is a `ref readonly` accessor if you want to modify you are only able to use the returned ref that comes from the `ref .Modify()` method.

File: src/Voron/Data/BTrees/TreeMutableState.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/TreeMutableState.cs:13]: We dont really win anything with the properties and force the JIT to have to reason about the access references. Having the ability to access the values from a reference makes the semantics explicit. First, the constraint of the `readonly ref` is enforced by the language and the JIT can use that knowledge to optimize (that is not true when using properties). Second, tracking places where modifications happen is as easy as do a reverse reference search over `.Modify()`. 

After having done this change and see what happens when we apply this methodology I would argue we should use more of this pattern, and get rid as much as possible of the track modification when executing property setters. Being explicit allow us to avoid common errors.

Comment (ayende) [reply src/Voron/Data/BTrees/TreeMutableState.cs:13]: Missed the `readonly` aspect, ok

File: src/Voron/Data/BTrees/Tree.Compressed.cs
Comment (ayende) [reply src/Voron/Data/BTrees/Tree.Compressed.cs:250]: I like this pattern

Comment (arekpalinski) [reply src/Voron/Data/BTrees/Tree.Compressed.cs:250]: Okay, I see. There is no way to do `State.NumberOfEntries--` directly. You are forced do call `State.Modify()` first. That's great.

**Comment by arekpalinski** [comment]
test this in debug please

File: src/Voron/Data/BTrees/Tree.Compressed.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/Tree.Compressed.cs:250]: When I discovered that I could do that was a WTF moment :D

File: src/Voron/Data/BTrees/Tree.Stream.cs
Comment (maciejaszyk) [inline src/Voron/Data/BTrees/Tree.Stream.cs:205]: Previously we've not set state as modified, this is a fix now?

File: src/Voron/Data/BTrees/Tree.cs
Comment (maciejaszyk) [inline src/Voron/Data/BTrees/Tree.cs]: In https://github.com/ravendb/ravendb/pull/18159, not a long time ago we removed disposable interface from trees since it can lead to double disposal. Is this safe? I see only usage is internal and by llt, maybe it's better to introduce some custom interface like IPageCachceReleaser just not to be confused with `IDisposable`. 

For me, when I see `IDisposable` that is clear that I have the responsibility of disposal but here is not the case, right?

Comment (ayende) [reply src/Voron/Data/BTrees/Tree.cs]: You cannot call `Tree.Dispose` directly, which helps.
But I agree about the intent

File: src/Voron/Data/BTrees/Tree.Stream.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/Tree.Stream.cs:205]: It was already set as modified before entering here. This is what I mean about being more explicit.

File: src/Voron/Data/BTrees/Tree.cs
Comment (redknightlois) [reply src/Voron/Data/BTrees/Tree.cs]: That's the reason why it implements `IDisposable` explicitly.

Comment (maciejaszyk) [reply src/Voron/Data/BTrees/Tree.cs]: Yes, but you can write
`using Tree tree = tx.CreateTree("tree");`

without any compiler exceptions.

**Comment by arekpalinski** [comment]
Please check tests failing in Debug:
https://jenkins.hibernatingrhinos.com/job/PR_Tests_Debug/97/

**Comment by redknightlois** [comment]
> Please check tests failing in Debug:
https://jenkins.hibernatingrhinos.com/job/PR_Tests_Debug/97/

This is an interesting issue. In fact the debug statement is correct, but the usage pattern is wrong because the API is highly inconsistent. I will set it up in such a way that we can avoid triggering it, but we need to solve the underlying problem.
https://issues.hibernatingrhinos.com/issue/RavenDB-22261/Inconsistency-in-Tree-external-API

**Comment by arekpalinski** [comment]
test this in debug please

**Comment by arekpalinski** [comment]
There are still Debug.Assert failures: https://jenkins.hibernatingrhinos.com/job/PR_Tests_Debug/99/testReport/

**Comment by redknightlois** [comment]
test this in debug please

**Comment by redknightlois** [comment]
If the pattern to avoid calling directly the `IDisposable` for `Tree` causes too many allocations I will revert it and write the code in such a way that if we ever use `using tree = new Tree(...)` statement it will not compile in `DEBUG` mode. Unless we want to introduce that right now. cc @ayende your call.

**Comment by arekpalinski** [comment]
test this in debug please

**Comment by arekpalinski** [comment]
Debug tests seems to be much better - https://jenkins.hibernatingrhinos.com/job/PR_Tests_Debug/106/testReport/. Can you please do the rebase (failing RavenDB_22176 tests got fixed meanwhile)?

**Comment by arekpalinski** [comment]
test this in debug please

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**
